### PR TITLE
 Graph tooltip, display and interval enhancements 

### PR DESCRIPTION
**Pull request by CDuffers**
At a high level, changes as follows:
1. Change tooltip to show give context related info, e.g. for a daily stat only show YMD
image  <-- Daily
image  <-- Larger than 24hrs
image  <-- less than 24hrs

2. Add support for touch devices within flot
3. Change the Interval to display and accept human readable figures, e.g. 1d, 2w, 4h
image 
4. Set the minTickSize to align with the YMD intervals
5. bar position is placed to the right for 'delta' to help show what time duration this covers
image   <-- With delta selected
image    <-- Without delta selected
6. Store the requesttype, e.g. fixed, daily, weekly ...
7. Align the start/ end dates to show the entire bar

image

### CODE REVIEW

**Comment by CDuffers** [comment]
The visual changes for this commit are in graph.js:
Setup a means to determine the language set within the browser which is set upon JS initiation.
Change the date/ time to use the browser language to display in the users locale.
Also applied the same for displaying numbers, to support the different decimal places and separators used. I’ve also tried to make the code a little easier to read, whilst also adhering to the decimal place request.

image <-- Browser set to en-gb
image <-- Browser set to de-de
image <-- Browser set to en-us

**Comment by CDuffers** [comment]
Hi Glyn,

Sorry, a poor description on my part.

And you're now making me think which way to cover this info anyway.

My approach was to differentiate the tooltip when the 'delta' option was selected. e.g. on a kWh feed this would give me ## kWh over a chosen duration (e.g hour, day) and hence this should show the start and end of that period.

This is where I'm interested in your thoughts: I'd initially considered the power type feeds as being purely a point in time, hence should only ever have a single date/ time.

However, thinking about it now, I presume the reality is that they are some form of sum of the previous 10 seconds since the last stat was collected, e.g. in your graph above, it wasn't that your usage was 2,178 at 21:22, but that the total for the previous 10 seconds was 2,178.

In short, if it is a sum over that period, then potentially all should show start and end. Doing so would remove a few if statements from the code and is easy to change.

image

The other changes add HH:MM:SS when less than a day and if 1 day, prevent it showing the same date twice

**Comment by pb66** [comment]
@glynhudson - Can we get this new feature merged into a new feature branch so we can test without adding new remotes. This should be the preferred method if commits are possibly incremental and/or subject to change based on feedback. I do not have access to this repo, do you @Paul-Reed ?

**Comment by Paul-Reed** [comment]
Sorry @pb66 - not at home right now.
Paul

**Comment by pb66** [comment]
Due to a recent clearout of my dev instance of emoncms I had no data to view, I set something up earlier and now I only have a couple of hours so my reviews are going to be somewhat limited for a day or two.

The first thing I noticed is the ugly float resolution has been reduced from

image

to this

image

IMO much better but for me it's been cut back a little to much, I would generally like to see 2 digits, although when viewing KW's I prefer 3 digits. 

The addition of seconds is great too! But I would prefer the 24hr clock was used and we lost the AM/PM, I like the order of month - date - hour - min - secs but then a return to 12 resolution at the end doesn't "fit" IMO.

Here we see 2 screen shots that show the same energy value in the hover over

image

image

in fact every result in this graph was either 0 or 0.1 kWh in the hoverover

image

I really like the from/to times so we know exactly what we are looking at, although it might be better to stack the from and to date/times so it is easier to spot the difference and recognize the interval.

It would also be nice to get a unit in the hover over too. I have raised a [feature request issue on the emoncms repo](https://github.com/emoncms/emoncms/issues/787) to see if we can get a unit included in the feeds table.

**Comment by CDuffers** [comment]
Hi Paul,

Thanks for the review.

**Date stacking,** like this?
image

**Number of decimal places** - To change, update the DP field in the input setup beneath the graph and it will show that many decimal places and allows for different DP per series.

The logic works by setting a default for the num DP as the feed is added to the graph. If the feed is a datatype of 2, or the first data point it finds is an integer (not fool proof, but a reasonable guide), then it sets the DP to 0. Otherwise it falls back to the previous default of '1'.

I can change this to 2 or 3 as collective preference suggests.

I've changed my code to default to 2 DP as below, yours should currently show 1 by default:
image

Which shows as:
image

A by-product of .toLocalString() is that it strips off trailing zeros, hence a point within the same graph and settings only shows 1 DP as the value is 0.10:
image

These values look a little better when you scale out the interval to hours, days etc


**AM/ PM** - Please check the language that your browser uses. I've found that if this is en-us, then it displays month before day and also a 12 hr clock, whilst en-gb is day, month and 24hr. I included examples within notes before Glyn's review.

This is all a byproduct of supporting localisation and is set within the .toLocaleString() and .toLocaleDateString() functions on both the display of the number (if you choose a German locale it uses ',' for a decimal place and '.' as thousands separator, you could try Chinese if you really want to bamboozle yourself, although I haven't) and the date/ time field.

**Units:** Yep, this is an area that I'd also thought was missing and would be very easy to add here. I'm happy to have a tinker in the future, but would like to get a couple of other bits done first.

As a potentially quirky idea, should your enhancement request also look to collate the preferred DP, then we can default to this, rather than have anything guesstimated?

**Comment by Paul-Reed** [comment]
In the dashboard widgets, the number of decimal point displayed is dependent upon the size of the number, ie;
```
        if (decimals<0){
            if (raw_value>=100) {
            raw_value = raw_value.toFixed(0);
            } else if (raw_value>=10) {
            raw_value = raw_value.toFixed(1);
            } else  {
            raw_value = raw_value.toFixed(2);
            }
         raw_value = parseFloat(raw_value);
         }
```
https://github.com/emoncms/dashboard/blob/master/widget/bar/bar_render.js#L179-L188
Could something along those lines be considered?

Paul

**Comment by CDuffers** [comment]
Hi Paul,

Yes, sure, could do. But as we have the DP field as a specified input, is there a reason not to use it?

A by product of the flexible approach is that as a user you can't get 3 DP.

It doesn't align to the thinking of whoever created the module as they'd specifically added the DP column, even if this only worked for the CSV output initially.

My line of thinking is if you specify for one, then it makes sense to keep it consistent with the others.

Did you have any thoughts with respect to start/ end times where a delta is NOT selected? i.e. should it remain as the time returned?

Clive.

**Comment by pb66** [comment]
_"Date stacking, like this?"_

That's what i was thinking, but that's just my opinion, what do you think?

_"Please check the language that your browser uses"_

Well I never....
I did not realise my browser was not following my OS settings, My OS is set to UK-English, but my browser was in fact set to "English", but I found both "UK-English" and "US-English" (and many others) were available options, setting UK-English has indeed given me dd-mmm, 24:00:00. Live and learn :-)

Is the comma thousands separator part of the same localisation or is that optional? I would opt for no thousands separator if it was an option since separators are not used in the scale values, on a bust graph 3,087 could be mistaken for 3.087, (my eyes ain't getting any better)

image

The DP setting seems to work to, great. I had no idea this related to the hover-over, I had thought it was just for the CSV output.

I think the DP is better definable rather than based on value size as you would have no DP on most accumulating feeds after a point. I'm not sure about putting DP in the feed data. I'm not sure it would be used anywhere else and it might not be obvious what it's for once it's out of context eg defining it on the feeds page.

_"Did you have any thoughts with respect to start/ end times where a delta is NOT selected? i.e. should it remain as the time returned?"_ 

If looking at an accumulating feed without delta selected then I think the value relates to a specific point in time (effectively the "end" time), the start time would effectively be the feed creation time so probably best left as a single time IMO.

**Comment by CDuffers** [comment]
Date stacking: It gets my vote

Browser language: Mine was similar as the default in Firefox was setting en-en.

Separators: I've looked and found a "useGrouping" option that can be specified. Personally, I prefer to have them for the graph types that I use, hence suggest allowing this as a user selectable option. I'm happy to add if everyone agrees.

If we do add this as an option then I think it worth adding an 'extended' toggle button to prevent bewildering new starters, with this in mind, I suggest grouping as follows:

**Basic:**
Feed
Type
Colour
Fill
Stack
Delta

**Extended:**
Numeric separators (New option for above)
Scale
Average
DP
Histogram

We can also save this setting with the graph.

In relation to the DP affecting the graph, I wonder whether we need some more help or guidance with this page as I've seen others question what 'DP' is in the community. Would adding some form of pop-up guidance make sense? We could restrict to just the headers, or also each series.

DP in the feed definition: My thinking was about adding it as an option, like your suggestion for units, rather than attempting to change source collection or equivalent.

I've not looked, but assume we could have an equivalent of feedid=##, unit="kWh", DP="2". So that this and other components would read the DP setting as a default and the user can override if they desire. It might then be handy for other widgets etc to pull this same info.

Start/ End times without Delta: Understood for accumulating feeds, does that apply to power feeds as well?

Clive.

**Comment by CDuffers** [comment]
Hi All,

I've started to tinker with the idea of a separator option and also standard/ extended options, here's a first attempt with "standard":
image

This view looks rather bland with big and really uneven gaps, hence wonder if allowing the table to resize would help. Any thoughts?

And with "extended" selected:
image

Any thoughts on placement of the selector? I wondered if I could fit it within the header bar of the feed table, between 'feeds in view' and 'show stats'.

And the separator field: should this be feed specific as per the screenshot or an option that applies to all?
